Minimum execution time for `withRootKey` on authentication failure.

jonathanknowles opened the pull request:
# Issue Number

#1984 

# Overview

This PR adjusts the `withRootKey` function to incorporate a minimum execution time in the event of failure.

- In the event of a _successful_ authentication, there is _no_ delay.
- In the event of an _unsuccessful_ authentication, the total execution time of the `withRootKey` function is:
    -  equal to (or greater than) `minimumExecutionTimeOnAuthFailure`.
    -  independent of the actual time required to perform the authentication.

Where `minimumExecutionTimeOnAuthFailure` is defined as:
```hs
-- | Specifies a minimum bound on the length of time that is allowed to elapse
--   before an authentication failure can be reported to the user.
--
minimumExecutionTimeOnAuthFailure :: NominalDiffTime   
minimumExecutionTimeOnAuthFailure = threeSeconds   
  where   
    threeSeconds = fromRational 3   
```

In the future, we could make this configurable (if the need arises).

# Testing

- [x] Add integration test to confirm that an unsuccessful attempt to authenticate results in a delay of `minimumExecutionTimeOnAuthFailure` or longer.
- [x] Add integration test to confirm that a successful attempt to authenticate does not result in such a delay.

jonathanknowles committed:
Add function `nominalDiffTimeToMicroseconds`.

jonathanknowles committed:
Add function `waitUntil`.

jonathanknowles committed:
Add function `withMinimumExecutionTimeOnFailure`.

jonathanknowles committed:
Add minimum execution time on authentication failure to function `withRootKey`.

jonathanknowles committed:
Introduce value `timeToExit`.

jonathanknowles committed:
Add test to confirm delay for unsuccessful authentications.

Add integration test to confirm that an unsuccessful attempt to
authenticate results in a delay of minimumAuthenticationFailureTime or
longer.

jonathanknowles committed:
Add test to confirm NO delay for successful authentications.

rvl commented on lib/core/src/Data/Time/Utils.hs:
```diff
@@ -23,3 +37,16 @@ utcTimeSucc = addUTCTime $ succ 0
+waitUntil :: MonadIO m => UTCTime -> m ()
+waitUntil futureTime = liftIO $ do
+    currentTime <- getCurrentTime
+    threadDelay
```
Does this work OK when the difference is negative?
Perhaps you could put a `max 0` there to make sure.

jonathanknowles replied:
> Does this work OK when the difference is negative?
> Perhaps you could put a `max 0` there to make sure.

I think it should work, as informal testing of `threadDelay` shows that it exits immediately when called with a negative value.

However, this behaviour appears to be undocumented, and so I agree it's a good idea to use `max 0`, as then we won't have to rely on undocumented behaviour.

Fixed in bcdeeb8d8327d9b91aacbd6ec71e5545547c4989.

rvl reviewed (commented):
It looks OK at first glance if we are going to implement this.

jonathanknowles committed:
Do not attempt to wait for a negative length of time.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1985#discussion_r468319927
